README.md
## wcag color contrast

[![CircleCI](https://circleci.com/gh/tmcw/wcag-contrast.svg?style=svg)](https://circleci.com/gh/tmcw/wcag-contrast)

This module produces [WCAG](http://www.w3.org/WAI/intro/wcag.php) [contrast ratio](http://www.w3.org/TR/WCAG20/#contrast-ratiodef)
measurement and scoring. It's an accessibility standard, intended to make sure websites have high enough contrast to be readable by everyone, including people with [low vision](https://www.aao.org/eye-health/diseases/low-vision) and age-related vision problems. This module is a low-level utility, useful for its implementation of the underlying math: to test a page, you'll want to use something like Google Chrome’s [Lighthouse tools](https://developers.google.com/web/tools/lighthouse/), or the [axe extension](https://www.deque.com/axe/) for Chrome and Firefox.

I (personally, speaking as Tom) have doubts about the [universality, accuracy, and application of this particular standard](https://macwright.org/2019/03/30/color-contrast-is-a-problem.html). High contrast is a good value, but I argue that the standard is over-strict and doesn't precisely match human perception.

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

### luminance

Get the contrast ratio between two relative luminance values

**Parameters**

-   `a` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** luminance value
-   `b` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** luminance value

**Examples**

```javascript
luminance(1, 1); // = 1
```

Returns **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** contrast ratio

### rgb

Get a score for the contrast between two colors as rgb triplets

**Parameters**

-   `a` **[array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)** 
-   `b` **[array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)** 

**Examples**

```javascript
rgb([0, 0, 0], [255, 255, 255]); // = 21
```

Returns **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** contrast ratio

### hex

Get a score for the contrast between two colors as hex strings

**Parameters**

-   `a` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** hex value
-   `b` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** hex value

**Examples**

```javascript
hex('#000', '#fff'); // = 21
```

Returns **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** contrast ratio

### score

Get a textual score from a numeric contrast value

**Parameters**

-   `contrast` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** 

**Examples**

```javascript
score(10); // = 'AAA'
```

Returns **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** score


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [3.0.0](https://github.com/tmcw/wcag-contrast/compare/v2.1.2-0...v3.0.0) (2019-11-05)


### ⚠ BREAKING CHANGES

* alters the output of failing tests. If you're using the empty string as a failure status, please adjust to looking for 'Fail' instead.

This adds `AA Large` from 3-4.5 and returns `'Fail'` when under 3 now. The empty string I think could be misconstrued as an error, since the output is not explicit.

### Features

* Support AA Large and Fail ([#57](https://github.com/tmcw/wcag-contrast/issues/57)) ([9dccf2a](https://github.com/tmcw/wcag-contrast/commit/9dccf2a24f9e20b9748379cb6d57970c76d5e844))

<a name="2.1.2-0"></a>
## [2.1.2-0](https://github.com/tmcw/wcag-contrast/compare/v2.1.1...v2.1.2-0) (2018-09-24)



<a name="2.1.1"></a>
## [2.1.1](https://github.com/tmcw/wcag-contrast/compare/v2.1.0...v2.1.1) (2018-09-24)



<a name="2.1.0"></a>
# [2.1.0](https://github.com/tmcw/wcag-contrast/compare/v2.0.0...v2.1.0) (2018-08-20)


### Features

* unpkg field ([eb9e840](https://github.com/tmcw/wcag-contrast/commit/eb9e840))



<a name="2.0.0"></a>
# [2.0.0](https://github.com/tmcw/wcag-contrast/compare/v1.2.0...v2.0.0) (2018-08-20)


### Features

* use microbundle, switch from tap to tape, improve structure ([3db73e2](https://github.com/tmcw/wcag-contrast/commit/3db73e2))


### BREAKING CHANGES

* new entry points, dependencies now bundled.



<a name="1.2.0"></a>
# [1.2.0](https://github.com/tmcw/wcag-contrast/compare/v1.0.0...v1.2.0) (2018-04-16)


### Bug Fixes

* Move esm dependency out of devDeps ([26a1553](https://github.com/tmcw/wcag-contrast/commit/26a1553))


### Features

* Internalize hexRgb dependency, making this and its dependents all es6-safe ([def207c](https://github.com/tmcw/wcag-contrast/commit/def207c))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/tmcw/wcag-contrast/compare/v1.0.0...v1.1.0) (2018-03-21)


### Features

* Internalize hexRgb dependency, making this and its dependents all es6-safe ([95a16e9](https://github.com/tmcw/wcag-contrast/commit/95a16e9))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/tmcw/wcag-contrast/compare/v0.1.1...v1.0.0) (2018-03-21)


### Bug Fixes

* **package:** update relative-luminance to version 1.0.0 ([59d9229](https://github.com/tmcw/wcag-contrast/commit/59d9229))



<a name="0.1.2-0"></a>
## [0.1.2-0](https://github.com/tmcw/wcag-contrast/compare/v0.1.1...v0.1.2-0) (2018-03-21)



<a name="0.1.1"></a>
## [0.1.1](https://github.com/tmcw/wcag-contrast/compare/v0.1.0...v0.1.1) (2018-01-10)


### Bug Fixes

* **package:** update hex-rgb to version 2.0.0 ([2ef3eb7](https://github.com/tmcw/wcag-contrast/commit/2ef3eb7))
* **package:** update relative-luminance to version 0.0.1 ([497542e](https://github.com/tmcw/wcag-contrast/commit/497542e))



# 0.1.0

* Add `hex(a, b)` method


test/contrast.test.js
const test = require("tape");
const c = require("../");

test("contrast", t => {
  t.equal(1, 1);
  t.end();
});

test("#score", t => {
  t.equal(c.score(0), "Fail", "zero contrast");
  t.equal(c.score(10), "AAA", "AAA");
  t.equal(c.score(5), "AA", "AA");
  t.end();
});

test("#rgb", t => {
  const white = [255, 255, 255];
  const black = [0, 0, 0];
  t.equal(c.rgb(white, white), 1, "same color");
  t.equal(c.rgb(white, black), 21, "same color");
  t.end();
});

test("#luminance", t => {
  t.equal(c.luminance(1, 1), 1, "same color");
  t.equal(c.luminance(1, 0), 21, "opposite luminance");
  t.end();
});

test("#hex", t => {
  const whiteHex = "#ffffff";
  const blackHex = "#000000";
  t.equal(c.hex(whiteHex, whiteHex), 1, "same color");
  t.equal(c.hex(whiteHex, blackHex), 21, "opposite luminance");
  t.end();
});


index.mjs
import relativeLuminance from "relative-luminance";

const hexChars = "a-f\\d";
const match3or4Hex = `#?[${hexChars}]{3}[${hexChars}]?`;
const match6or8Hex = `#?[${hexChars}]{6}([${hexChars}]{2})?`;

function hexRgb(hex) {
  hex = hex.replace(/^#/, "");
  let alpha = 255;

  if (hex.length === 8) {
    alpha = parseInt(hex.slice(6, 8), 16);
    hex = hex.substring(0, 6);
  }

  if (hex.length === 4) {
    alpha = parseInt(hex.slice(3, 4).repeat(2), 16);
    hex = hex.substring(0, 3);
  }

  if (hex.length === 3) {
    hex = hex[0] + hex[0] + hex[1] + hex[1] + hex[2] + hex[2];
  }

  const num = parseInt(hex, 16);
  const red = num >> 16;
  const green = (num >> 8) & 255;
  const blue = num & 255;

  return [red, green, blue, alpha];
}

// http://www.w3.org/TR/WCAG20/#contrast-ratiodef

/**
 * Get the contrast ratio between two relative luminance values
 * @param {number} a luminance value
 * @param {number} b luminance value
 * @returns {number} contrast ratio
 * @example
 * luminance(1, 1); // = 1
 */
export function luminance(a, b) {
  const l1 = Math.max(a, b);
  const l2 = Math.min(a, b);
  return (l1 + 0.05) / (l2 + 0.05);
}

/**
 * Get a score for the contrast between two colors as rgb triplets
 * @param {array} a
 * @param {array} b
 * @returns {number} contrast ratio
 * @example
 * rgb([0, 0, 0], [255, 255, 255]); // = 21
 */
export function rgb(a, b) {
  return luminance(relativeLuminance(a), relativeLuminance(b));
}

/**
 * Get a score for the contrast between two colors as hex strings
 * @param {string} a hex value
 * @param {string} b hex value
 * @returns {number} contrast ratio
 * @example
 * hex('#000', '#fff'); // = 21
 */
export function hex(a, b) {
  return rgb(hexRgb(a, { format: "array" }), hexRgb(b, { format: "array" }));
}

/**
 * Get a textual score from a numeric contrast value
 * @param {number} contrast
 * @returns {string} score
 * @example
 * score(10); // = 'AAA'
 */
export function score(contrast) {
  return contrast >= 7 ? "AAA" : contrast >= 4.5 ? "AA" : "";
}


index.js
import relativeLuminance from "relative-luminance";

const hexChars = "a-f\\d";
const match3or4Hex = `#?[${hexChars}]{3}[${hexChars}]?`;
const match6or8Hex = `#?[${hexChars}]{6}([${hexChars}]{2})?`;

function hexRgb(hex) {
  hex = hex.replace(/^#/, "");
  let alpha = 255;

  if (hex.length === 8) {
    alpha = parseInt(hex.slice(6, 8), 16);
    hex = hex.substring(0, 6);
  }

  if (hex.length === 4) {
    alpha = parseInt(hex.slice(3, 4).repeat(2), 16);
    hex = hex.substring(0, 3);
  }

  if (hex.length === 3) {
    hex = hex[0] + hex[0] + hex[1] + hex[1] + hex[2] + hex[2];
  }

  const num = parseInt(hex, 16);
  const red = num >> 16;
  const green = (num >> 8) & 255;
  const blue = num & 255;

  return [red, green, blue, alpha];
}

// http://www.w3.org/TR/WCAG20/#contrast-ratiodef

/**
 * Get the contrast ratio between two relative luminance values
 * @param {number} a luminance value
 * @param {number} b luminance value
 * @returns {number} contrast ratio
 * @example
 * luminance(1, 1); // = 1
 */
export function luminance(a, b) {
  const l1 = Math.max(a, b);
  const l2 = Math.min(a, b);
  return (l1 + 0.05) / (l2 + 0.05);
}

/**
 * Get a score for the contrast between two colors as rgb triplets
 * @param {array} a
 * @param {array} b
 * @returns {number} contrast ratio
 * @example
 * rgb([0, 0, 0], [255, 255, 255]); // = 21
 */
export function rgb(a, b) {
  return luminance(relativeLuminance(a), relativeLuminance(b));
}

/**
 * Get a score for the contrast between two colors as hex strings
 * @param {string} a hex value
 * @param {string} b hex value
 * @returns {number} contrast ratio
 * @example
 * hex('#000', '#fff'); // = 21
 */
export function hex(a, b) {
  return rgb(hexRgb(a, { format: "array" }), hexRgb(b, { format: "array" }));
}

/**
 * Get a textual score from a numeric contrast value
 * @param {number} contrast
 * @returns {string} score
 * @example
 * score(10); // = 'AAA'
 */
export function score(contrast) {
  if (contrast  >= 7) {
    return 'AAA';
  }
  
  if (contrast >= 4.5) {
    return 'AA';
  }
  
  if (contrast >= 3) {
    return 'AA Large';
  }
  
  return 'Fail';
}
